### DISCUSSION TITLE
How does a tool with several scripts should be handled in manifest ?

### DISCUSSION THREAD

**Discussion by cderv** [ScoopInstaller/Scoop#5156] [Help and Questions] [answered]
I am trying to improve pyenv manifest: https://github.com/ScoopInstaller/Main/blob/master/bucket/pyenv.json

pyenv-win project already contains some scripts to use in powerwhell, CMD and in Git bash. 
````
> ls "$(scoop prefix pyenv)/pyenv-win/bin"

    Directory: C:\Users\chris\scoop\apps\pyenv\current\pyenv-win\bin

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d----          20/07/2022    16:27                WiX
-----          20/07/2022    16:27            151 pyenv
-----          20/07/2022    16:27           6885 pyenv.bat
-----          20/07/2022    16:27            611 pyenv.ps1
````

I would think that we should make those available as scoop shims for better handling the different 

However, trying to use the `bin` field in manifest, I don't get expected result
````
    "bin": [
        "pyenv-win\\bin\\pyenv.bat",
        "pyenv-win\\bin\\pyenv.ps1",
        "pyenv-win\\bin\\pyenv"
    ],
````

In scoop shims: 
* `pyenv.ps1` will correctly be a wrapper to call `pyenv-win\\bin\\pyenv.ps1`
````powershell
───────┬────────────────────────────────────────────────────────────────────────────────────────────────────────────────
       │ File: pyenv.ps1
───────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1   │ # C:\Users\chris\scoop\apps\pyenv\current\pyenv-win\bin\pyenv.ps1
   2   │ $path = Join-Path $PSScriptRoot "..\apps\pyenv\current\pyenv-win\bin\pyenv.ps1"
   3   │ if ($MyInvocation.ExpectingInput) { $input | & $path  @args } else { & $path  @args }
   4   │ exit $LASTEXITCODE
````
* `pyenv` will be a wrapper to call the `pyenv-win\\bin\\pyenv` 
````bash
───────┬────────────────────────────────────────────────────────────────────────────────────────────────────────────────
       │ File: pyenv
───────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1   │ #!/bin/sh
   2   │ # C:\Users\chris\scoop\apps\pyenv\current\pyenv-win\bin\pyenv
   3   │ "C:\Users\chris\scoop\apps\pyenv\current\pyenv-win\bin\pyenv"  "$@"
````

* But `pyenv.cmd` will not be a wrapper to call `pyenv.bat`, it will be another way to call `pyenv` using GIt bash
````cmd
───────┬────────────────────────────────────────────────────────────────────────────────────────────────────────────────
       │ File: pyenv.cmd
───────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1   │ @rem C:\Users\chris\scoop\apps\pyenv\current\pyenv-win\bin\pyenv
   2   │ @"C:\Program Files\Git\bin\bash.exe" "C:\Users\chris\scoop\apps\pyenv\current\pyenv-win\bin\pyenv"  %*
````

Is this really expected ? 

I am seeing in the code that 

* `.cmd` is created in bin is a `bat`
https://github.com/ScoopInstaller/Scoop/blob/4a31bd330244f7f89f16208cdddda3f9edac2d65/lib/core.ps1#L727-L740
* but also if it is a file without extension,
https://github.com/ScoopInstaller/Scoop/blob/4a31bd330244f7f89f16208cdddda3f9edac2d65/lib/core.ps1#L811-L827

It seems there will always be overwrite not matter the order I am putting for `bin`.  For example putting `.ps1 last will create the three scoop shims based running `pyenv-win\\bin\\pyenv.ps1` in powershell

How can I set the fields to account for the tools here already creating script of Powershell, CMD and Git-bash  shells ? 

Workaround I found for now would be 

* To not set any scoop shims - so no `bin` field in manifest
* Add the `pyenv-win\\bin` folder to PATH, as recommended in their [setup instruction](https://github.com/pyenv-win/pyenv-win#add-system-settings)
````
    "env_add_path": ["pyenv-win\\bin", "pyenv-win\\shims"],
````

Is that supported, is this something that would do it ? 

Thank you

**Comment by rashil2000** [comment, marked as answer]
If the package already provides scripts for different shells (i.e. bash, cmd, pwsh), then the recommended way is to use `env_add_path` and avoid `bin` field.

**Reply by cderv** [reply]
Thanks. That answers my question then , and confirm my understanding after digging through this. I really wasn't sure this was the recommended way. 

i'll make a PR to update pyenv manifest to support this better then. 

Thank you !

**Reply by cderv** [reply]
FWIW I opened PR to update the manifest: https://github.com/ScoopInstaller/Main/pull/3950
